lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1368,11 +1041,16 @@ def list_from_ruby(arr)
       ret
     end
 
-    def struct_from_ruby(hash)
+    # If reduce is true, nil will be returned for hash like
+    # {"httpRequest"=>{"referer"=>nil}} instead of layers of empty structs.
+    def struct_from_ruby(hash, reduce = true)
       ret = Google::Protobuf::Struct.new
       hash.each do |k, v|
-        ret.fields[k] ||= value_from_ruby(v)
+        ret.fields[k] ||= value_from_ruby(v) \
+          if !v.nil? && !value_from_ruby(v).nil?
```

Comment:
Did some investigation on how the REST path handles null fields. Turned out my previous implementation did break the existing functionality with null field handling. The previous fix has been reverted, and a new fix that directly applies to the `set_http_request_grpc` method has been included. And some new tests covering the null field handling behavior have been added.